### PR TITLE
Adjustments to the OML mixin and MetaContinualLearningExperiment class

### PR DESCRIPTION
**Pull request by mvacaporale**
This PR adjust and revises the meta-continual experiment codebase for a more faithful implementation of the OML algorithm.

The major changes include:
* An extraneous pass through the model is removed that prevented backproping through the adapted network (commit: [0faf8fd](https://github.com/numenta/nupic.research/pull/389/commits/0faf8fdd6ba1b919fa003beb8c689969b1a9267f))
* That change surfaced a bug, and now `update_params` works by accessing and updating the `_parameters` of the fast-param related modules (this is how the learn2learn repo does it). (commit: [dc16df2](https://github.com/numenta/nupic.research/pull/389/commits/dc16df2dfa5c2852293f312703615924accb7d4b))
* To support this change, the `OMLNetwork` now has attributes for `named_fast_params` and `named_slow_params` (instead of the unnamed version); these names help access the appropriate `_parameters` attribute of the model. (commit: [5e26602](https://github.com/numenta/nupic.research/pull/389/commits/5e2660244d6e737feccf18c6c8573a41c6732d09))
* Meta-train training now performs `num_fast_steps` within the inner loop. The OML algorithm depends on a sequence of gradient steps, not one gradient step over a batch of images. (commit: [8ad02c6](https://github.com/numenta/nupic.research/pull/389/commits/8ad02c6a8aba8b7ef5d20855854ed472a5675ef0))
* There's a `pre_task` method for `MetaContinualLearningExperiment`. For OML, this is helpful for resetting the output layer weights for the corresponding classes (i.e. tasks) about to be trained on. (commits: [6769a72](https://github.com/numenta/nupic.research/pull/389/commits/6769a72ed71a3dcecff94f7d358de8570d021e5e), [d627123](https://github.com/numenta/nupic.research/pull/389/commits/d6271234e6b458aaae83d3e7bbe60849de4cefe9))
* The meta-testing phase (within `OnlineMetaLearning`) now uses the appropriate loss function. Before it used the default set by `train_model` and `evaluate_model` which was `nll_loss`. Now it uses the one defined via the config: `self._loss_function` which is set to `cross_entropy` by default. (commit: [d00e593](https://github.com/numenta/nupic.research/pull/389/commits/d00e59391e5f3a1542c749881e081207ea508ae9))
* The meta-testing phases now learns over the sampled classes sequentially. Before, the samples were i.i.d and classes were mixed among one another. (commit: [df94d7e](https://github.com/numenta/nupic.research/pull/389/commits/df94d7ed5545ab02a180a4339f77fd95f1b1152d))
* The `OnlineMetaLearning` mixin now has it's own `create_*_loader` functions so it may define it's own train/test split. In the OML repo, the meta-testing phase trains over 15 images per class and tests over the remaining 5 images per class, whereas the meta-training phase does not use any such split. (commit: [b9d1d0e](https://github.com/numenta/nupic.research/pull/389/commits/b9d1d0e3e23f38d949564484c3b8da8b948acf29))
* There's a new config `metacl_oml_replicate` in the `meta_cl` project folder that adjusts the hyperparameters to more accurately replicate the OML experiment. There are still some differences between our implementation and theirs. These will be fleshed out in a JIRA and I'll tag you both. (commit: [f8d5d16](https://github.com/numenta/nupic.research/pull/389/commits/f8d5d16ef9ed87a30cc229977da71189345bcfe2))

### CODE REVIEW

**Comment by mvacaporale** [comment]
@lucasosouza @karangrewal To make the review process easier, I've listed the key changes in the PR description. As well, I've now added links to their corresponding commits. This way you can go straight to major changes instead of reading through all the commits (as I've suggested before).

File: nupic/research/frameworks/pytorch/models/common_models.py [hunk 1]
```diff
@@ -119,12 +119,30 @@ def conv_block(cls, in_channels, out_channels, kernel_size, stride, padding):
...
 
     @property
-    def fast_params(self):
-        return self.adaptation.parameters()
+    def named_fast_params(self):
+        """
+        Returns named params of adaption network, being sure to prepend
+        the names with "adaptation."
+        """
+        named_parameters = self.adaptation.named_parameters()
```
Comment (karangrewal) [inline nupic/research/frameworks/pytorch/models/common_models.py:127]: Why don't we replace `adaptation` and `representation` with `fast` and `slow`, respectively, everywhere? That may make things simpler.

File: nupic/research/frameworks/vernon/mixins/oml.py [hunk 2]
```diff
@@ -173,3 +280,31 @@ def get_execution_order(cls):
         eo["create_loaders"].append("Create loaders for the meta testing phase")
 
         return eo
+
+
+class ResetOMLTaskParams(object):
```
Comment (karangrewal) [inline nupic/research/frameworks/vernon/mixins/oml.py:285]: can you explain why we need to make this a separate mixin, rather than just defining the `pre_task` method used here in the OMLExperiment class itself?

Comment (mvacaporale) [reply nupic/research/frameworks/vernon/mixins/oml.py:285]: Fair point. As of now `ResetOMLTaskParams` only works with the `OMLNetwork` whereas the `OnlineMetaLearning` could technically work for any network and is thus broader in scope.

If there was a generalized way to reset the parameters as needed, then I think we could incorporate it into `OnlineMetaLearning`, but I'm not sure that's the case. For instance, even if it always involved resetting the weights in the last linear layer, how will we know how to access those weights? They typically will be `list(self.model.parameters())[-2]`, but that will not necessarily always be true.

We could also enforce the model to have an `output_params` attribute, but I didn't want to enforce more constraints past the `named_fast_params` and `named_slow_params`. At any rate, I figure the best solution will come with further experiments and further understanding our needs.

File: nupic/research/frameworks/pytorch/models/common_models.py (hunk 1 above)
Comment (mvacaporale) [reply nupic/research/frameworks/pytorch/models/common_models.py:127]: The `fast` and `slow` scheme was helpful for us in defining the `MetaContinualLearningExperiment`, but I'm not sure it makes sense to apply it broadly. Here, the `OMLNetwork` is implemented to mimic that of the papers. As one reads the code, they can make the connection mentally to the paper.

In contrast, `named_fast_params` and `named_slow_params` serves as an interface and helps the model interact with the experiment class.

Note, we'll always need such an interface. It helps keep the experiment agnostic to the model. Imagine if we have a network with multiple dendritic segments. We may not be able to divide our modules cleanly into `self.slow` and `self.fast`. Thus, we can allow the model to be defined in whatever way it needs to be, as long as it includes the interface.

Comment (mvacaporale) [reply nupic/research/frameworks/pytorch/models/common_models.py:127]: I see your point though `named_fast_params` is a bit cumbersome. I can think of a couple of other ways to do it, but I'm not sure they're any better.

File: nupic/research/frameworks/vernon/mixins/oml.py [hunk 3]
```diff
@@ -103,13 +102,15 @@ def find_best_lr(self, num_classes_learned):
...
 
                 optim = Adam(params, lr=lr),
-                train_model(
-                    model=self.model.module,
-                    loader=self.test_train_loader,
-                    optimizer=optim,
-                    device=self.device,
-                    criterion=self._loss_function,
-                )
+                for task in new_tasks:
```
Comment (lucasosouza) [inline nupic/research/frameworks/vernon/mixins/oml.py:169]: This seems to be more accurate. But just as a sanity check, is this how the OML repo is doing? Because if I understood correctly from the first PR the whole meta testing phase was a copy and paste from OML with the few adaptations required to match vernon structure.

File: nupic/research/frameworks/vernon/handlers.py [hunk 4]
```diff
@@ -1311,15 +1313,20 @@ def update_params(cls, params, loss, lr, distributed=False):
...
 
         if gradients is not None:
-            params = list(params)
-            for p, g in zip(params, gradients):
+            for g, (name, p) in zip(gradients, named_params.items()):
                 if g is not None:
-                    p.add_(g, alpha=-lr)
+                    updated = p.add(g, alpha=-lr)
+
+                    # Update in place inn a way that preserves grads.
```
Comment (lucasosouza) [inline nupic/research/frameworks/vernon/handlers.py]: small typo, inn

File: nupic/research/frameworks/vernon/handlers.py [hunk 5]
```diff
@@ -1311,15 +1313,20 @@ def update_params(cls, params, loss, lr, distributed=False):
...
+                    parent_module = get_parent_module(model, name)
+                    base_name = name.split(".")[-1]
+                    parent_module._parameters[base_name] = updated
```
Comment (lucasosouza) [inline nupic/research/frameworks/vernon/handlers.py]: I don't understand the bug, why can't it be done inplace?

File: nupic/research/frameworks/pytorch/models/common_models.py (hunk 1 above)
Comment (lucasosouza) [reply nupic/research/frameworks/pytorch/models/common_models.py:127]: I agree with Karan, keep the nomenclature stable across the framework. The link to adaptation / representation can be made clear in the docstring instead for potential readers/users.

But you certainly gave a lot more thought to this, so however you think it's best.

File: nupic/research/frameworks/vernon/handlers.py [hunk 6]
```diff
@@ -1133,6 +1135,7 @@ def setup_experiment(self, config):
         self.val_batch_size = config.get("val_batch_size", 15)
         self.slow_batch_size = config.get("slow_batch_size", 64)
         self.replay_batch_size = config.get("replay_batch_size", 64)
+        self.num_fast_steps = config.get("num_fast_steps", 1)
```
Comment (lucasosouza) [inline nupic/research/frameworks/vernon/handlers.py]: The name is a bit misleading, since it will only take as much as steps as the number of (samples in class / batch_size). To effectively take the number of steps requested by the user it would have to do as many loops through the loader as required, but that is not what is happening in the train function below. 

Maybe we can add a message stating the effective number of fast steps in case the number given by the user is greater than (samples in class / batch_size)

File: nupic/research/frameworks/vernon/handlers.py [hunk 7]
```diff
@@ -1278,14 +1281,17 @@ def run_task(self, task, cloned_adaptation_net):
         self.val_fast_loader.sampler.set_active_tasks(task)
 
         # Train, one batch
```
Comment (lucasosouza) [inline nupic/research/frameworks/vernon/handlers.py]: This is actually several batches of one sample, I don't believe the "one batch" comment is still accurate.

File: nupic/research/frameworks/vernon/handlers.py [hunk 8]
```diff
@@ -1282,6 +1275,14 @@ def run_epoch(self):
 
         return results
 
+    def pre_task(self, tasks):
```
Comment (lucasosouza) [inline nupic/research/frameworks/vernon/handlers.py]: Might be useful in Continual Learning too, we can add later.

File: nupic/research/frameworks/vernon/mixins/oml.py [hunk 9]
```diff
@@ -156,14 +158,16 @@ def run_meta_testing_phase(self, num_classes_learned):
...
+                criterion=self._loss_function,
             )
 
             # meta-testing testing
             results = evaluate_model(
                 model=self.model.module,
                 loader=self.test_test_loader,
-                device=self.device
+                device=self.device,
+                criterion=self._loss_function,
```
Comment (lucasosouza) [inline nupic/research/frameworks/vernon/mixins/oml.py:258]: Good catch!

File: nupic/research/frameworks/vernon/mixins/oml.py [hunk 10]
```diff
@@ -51,6 +53,9 @@ def setup_experiment(self, config):
                                   times, i.e. the mode, will be chosen for the
                                   meta-testing phase.
             - num_meta_testing_runs: number of meta-testing phases to run
+            - meta_test_sample_size: number of images per class to sample from for
```
Comment (lucasosouza) [inline nupic/research/frameworks/vernon/mixins/oml.py:56]: The naming doesn't make clear it is for training - shouldn't be meta_test_train_sample_size? I know it is a ridiculously long name, but we can drop the meta instead, just `test_train_sample_size`.

File: projects/meta_cl/experiments/oml.py [hunk 11]
```diff
@@ -79,8 +79,91 @@ def run_experiment(config):
...
+metacl_oml_replicate.update(
+
+    # How often to checkpoint (in epochs)
+    checkpoint_freq=1000,
+    keep_checkpoints_num=1,
+    checkpoint_at_end=True,
+    checkpoint_score_attr="training_iteration",
+
+    # Number of tasks (i.e. distinct classes) for the inner loops and outer loop
+    tasks_per_epoch=3,
```
Comment (lucasosouza) [inline projects/meta_cl/experiments/oml.py:94]: Is it just 3?

**Review by lucasosouza — APPROVED** [review]

**Comment by lucasosouza** [comment]
Looks good, left some comments, might be worth going through them. And a few questions.

**Review by karangrewal — APPROVED** [review]

File: nupic/research/frameworks/pytorch/models/common_models.py (hunk 1 above)
Comment (mvacaporale) [reply nupic/research/frameworks/pytorch/models/common_models.py:127]: I agree to keep the nomenclature consistent. However, I think it's important to have two separate classifications: one for our modules and one for the params. The `OMLNetwork` has `adaptation` which has all of the fast params and `representation` which has all the slow params, but maybe there won't always be a perfect overlap between the two. That's my concern. We can keep the naming in mind though as we move forward, and change it if we see it's too confusing.

File: nupic/research/frameworks/vernon/mixins/oml.py (hunk 3 above)
Comment (mvacaporale) [reply nupic/research/frameworks/vernon/mixins/oml.py:169]: It was copy-paste, but we had to translate the parts that dealt with iterating through the data. The OML repo didn't iterate task-by-task because they pre-ordered their dataset to iterate over the tasks in sequence (See this [line](https://github.com/khurramjaved96/mrcl/blob/2855a6b7e820f171432981b58c49664fcdbf00ed/evaluate_omniglot.py#L144)). In contrast, our samplers automatically shuffle the data while we iterate through the dataloader.

Note, in the future, I think it will be better to make sure our task samplers (such as `TaskDistributedSampler`) can preserve the order of the tasks. They technically can if you set `shuffle=False`, but I wondered about another case where we preserve the order of the tasks but then shuffle samples within each task. I left this till later to keep things simple for now.

File: nupic/research/frameworks/vernon/handlers.py (hunk 6 above)
Comment (mvacaporale) [reply nupic/research/frameworks/vernon/handlers.py]: Good point, I'll add a warning.

File: nupic/research/frameworks/vernon/handlers.py (hunk 7 above)
Comment (mvacaporale) [reply nupic/research/frameworks/vernon/handlers.py]: Good catch.

File: nupic/research/frameworks/vernon/mixins/oml.py (hunk 9 above)
Comment (mvacaporale) [reply nupic/research/frameworks/vernon/mixins/oml.py:258]: Thanks, this was an interesting one hunt down.

File: nupic/research/frameworks/vernon/mixins/oml.py (hunk 10 above)
Comment (mvacaporale) [reply nupic/research/frameworks/vernon/mixins/oml.py:56]: Good point. I'll update this name as well as `meta_train_sample_size` -> `train_train_sample_size`.

File: projects/meta_cl/experiments/oml.py (hunk 11 above)
Comment (mvacaporale) [reply projects/meta_cl/experiments/oml.py:94]: It is, this is taken from the OML readme where they suggest running the following to replicate their results:

```
python oml_omniglot.py --update_lr 0.03 --meta_lr 1e-4 --name OML_Omniglot/ --tasks 3 --update_step 5 --steps 700000 --rank 0
```

Notice `--tasks 3`. The results reported in `projects/meta_cl/experiments/oml.py` come from running OML's code with this configuration. So the comparison is fair in any case.

File: nupic/research/frameworks/vernon/handlers.py (hunk 5 above)
Comment (mvacaporale) [reply nupic/research/frameworks/vernon/handlers.py]: This is a good question and it deserves a good answer. I'm still trying to understand a bit more about why it's needed (other than the fact it fixed a bug PyTorch was throwing). Let me get back to you on this.

**Comment by mvacaporale** [comment]
> Looks good, left some comments, might be worth going through them. Had a few questions too.
> 
> That is a lot of commits =)

Thanks for the feedback. I might go easier on the commits next time 😅

File: nupic/research/frameworks/vernon/mixins/oml.py (hunk 3 above)
Comment (lucasosouza) [reply nupic/research/frameworks/vernon/mixins/oml.py:169]: That is a good observation. We definitely can distinguish between shuffling tasks and shuffling samples within tasks in the sampler.

File: projects/meta_cl/experiments/oml.py (hunk 11 above)
Comment (lucasosouza) [reply projects/meta_cl/experiments/oml.py:94]: I see. I don't remember seeing the 3 anywhere in the paper, but I think they just don't mention that part. For some reason I had 5 in my mind.
